### ISSUE TITLE
Implementing SMART APP Invocation for OpenEMR

### ISSUE THREAD

**Issue by vishnuyar** [openemr/openemr#4029]
### Describe the problem

OpenEMR currently has FHIR Server capability. To further enhance the functionality for OpenEMR, we should implement SMART APP launch based on this framework http://hl7.org/fhir/smart-app-launch/

### Describe the solution you'd like

Create a Form for SMART APP Launch 
This can be installed from Administration -->Forms --> Forms Administration

Details for individual smart app applications -- such as launch url and redirect uri can be entered in the form

Will need for further help on authentication. -- 

I can start implementing for this features once solution presented is approved.

**Comment by sjpadgett** [comment]
Hi @vishnuyar 
Not sure if you've been keeping up with my current project at https://community.open-emr.org/t/v6-authorization-and-api-changes-afoot/15450 but SMART is next on my todo.

If you can contribute towards that end it would be greatly appreciated.

**Comment by vishnuyar** [comment]
I was not aware of the project. Please let me know how i can contribute towards the same.

**Comment by sjpadgett** [comment]
Thanks @vishnuyar 
umm, where to start!
Currently my OIDC identity server has a broad range of scopes and claims that I need to trim to openemr specific uses. Scopes like email, phone and so forth are not needed because we already have patient/user consent to hold that data. For SMART I need to add more specific scopes dealing with what type of access is allowed such as patient/Observation.read or patient/Observation.write or launch and so on.

This is going to require a marriage with our resource server and thus, api dispatch to determine what to serve. I hope to do this with ACL but, by no means is the only way. A look up comes to mind. 
Anyway, I won't get to this for a week or two unless you have some ideas. 

Still if you wanted to start on your proposed SMART form/app you can authenticate to our OIDC/OA2 and hold off on any necessary SOF scopes needed until we can get the SMART delegation set up on servers work flow.

You prob know this stuff but, may as well do some documenting for others that are keeping up with project.

Let me know if questions

**Comment by vishnuyar** [comment]
@bradymiller  
how do i enable CORS for OpenEMR

I am currently getting the following error 

"Access to XMLHttpRequest at 'https://******/latest/apis/default/fhir/metadata' from origin 'https://*****' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.

**Comment by bradymiller** [comment]
hi @vishnuyar , Unable to confirm this issue on postman or browser. What is your environment/browser?

**Comment by vishnuyar** [comment]
I am using an oauth client, which will be the redirect_uri when registering the client. Before registering the client in OpenEMR, the oauth client will try to get the OpenEMR Auth Endpoint from the metadata via browser call. I am currently using latest google chrome.

**Comment by sjpadgett** [comment]
I think you need to turn cors off in browser or add a directive to your server Header set Access-Control-Allow-Origin "*" if apache
For AWS it's a different process that you'll have to look up.

I'm wondering if this isn't something we can config in our API.

Maybe @jesdynf can advise.

**Comment by jesdynf** [comment]
Browsers can't turn off CORS. I mean, you can almost do it with certain launch parameters and certain add-ons, but that's neither viable nor 100% effective. The real problem isn't ACAO * or the lack of it, it's the CORS preflight check that the browser is requiring. Baaaaasically what happens if you either use a non-standard HTTP verb, or you try to use any headers that aren't on a very short list, or you just basically try to do anything interesting, your browser will ask the target server if that's permitted via the CORS preflight check, an HTTP `OPTIONS` call that expects certain headers back.

https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

So our options are:
 * Take no action because we don't believe browsers *should* be using this API.
 * Implement preflight support in PHP. https://www.dinochiesa.net/?p=754
 * Implement preflight support in Apache. https://stackoverflow.com/questions/25709175/handling-cors-preflight-request-in-apache

You may want to accept or deny CORS requests based on the referrer or the host the request originates from, and since I'm not 100% sure the OpenEMR container knows its own hostname, it'd be perfectly reasonable to add these configuration chunks and leave them commented out, asking the administrator to unlock them themselves once they're ready to.

One point I'd like to make is that CORS isn't a security feature on its own -- CORS doesn't have the duty of blocking requests. CORS is a tool that helps *the browser* block requests, because you can tell the browser that only https://cooldudes.com should be using this API, so that if the browser's user is getting clickjacked by hostile code on https://scrublords.net, the browser knows that scrublords has no right to initiate requests to cooldudes. This is what the AC* headers and preflight checks are for, the browser deciding whether or not to complete requests or furnish data to the browser -- and this is why you *can't turn CORS off*, because exactly nobody wants to deal with this mess unless you make them.

**Comment by sjpadgett** [comment]
Thanks @jesdynf nicely put.
So, no longer can config around browser however, why would one want to if security is of import.
The only legit reason a browser may want to be used with our apis is in this case of grabbing meta or well knowns for download.
Having admin/it config server would be my last choice leaving I suppose, supporting the http option.

I'll look into.

**Comment by vishnuyar** [comment]
Based on the above, My understanding is
1) The Server running the OpenEMR should allow Cors and options as
mentioned in
https://stackoverflow.com/questions/25709175/handling-cors-preflight-request-in-apache
2) The application that is OpenEMR should give the admin the settings to
set Headers mentioned in https://www.dinochiesa.net/?p=754

Both are a must and satisfying one condition itself doesn't work.

**Comment by jesdynf** [comment]
Not truly. The requirement is met if the browser can make a preflight request and get a positive response. If the app attends to that need, you can get something fairly fine-grained and responsive, an arbitrarily complex PHP evaluation. If the server does it, which it can through rewriting, then you're going to get something less flexible but still reasonable, hardcoding the server's name into the access headers it returns.

So you could do it just with the Apache configs, if you were of a mind to.

**Comment by sjpadgett** [comment]
@vishnuyar 
Here is the issue for the server part for tracking progress and bringing up issues as you come across in your development: https://github.com/openemr/openemr/issues/4076

In regard to the CORS issue you and Asher @jesdynf have commented on, i'm not so sure we'd want the solution in server config  because I worry it would be to global to the whole OpenEMR enterprise instance. My thinking is that primarily dealing with conformance and/or any .well known endpoints is the job of the client thus implemented outside the browser. Though, I don't know how most vendors or app developers deal with this workflow.

Is it normal like say a SOAP wsdl retrieve could be internal or external for class building. Where with auth server or any api conformance i'd think it would most always be dynamic in the client and i'm guessing any browser GETs would be for testing and development purposes.

So bottom line to me is that this would really be a documentation informative rather than implementing a solution for production. Even here I'd think we need keep the focus on where a preflight response is needed i.e the endpoints for well known or conformance and implement a restrictive http options response in resource server. I'm not an IT guy so I may be overthinking this.

**Comment by sjpadgett** [comment]
@vishnuyar We have started the SMART layer for auth/identity server. @adunsulag is heading this up where you may want to interface with him. His starting focus is on the EMR launch.

So, are you still pursuing this app and if so, where do ya stand? Need help?

**Comment by vishnuyar** [comment]
@sjpadgett , I am struggling with CORS pre-flight check when invoking OpenEMR Oauth2 from the Smartapp callback url
I am writing headers in the apache config but haven't progressed much. 
@adunsulag has covered the security part of FHIR metadata which I was planning to push. So we are good there.

I am thinking, we will need a create a UI for registering  SMART App's. "This will invoke oauth2 registration. "
The UI will also have to provide location of menu link for this smart app to be invoked from within OpenEMR similar to Access Control in Forms Administration. This UI will also give the Administrator to de-register SMART Apps.

This whole UI will be a form module to to be enabled from Forms Administration

**Comment by sjpadgett** [comment]
@vishnuyar 
@adunsulag and I spoke about the CORS and agree(I think) we should implement handling pre-flight in our resource server via an http options response(PHP solution). Want to take a shot at doing this?
Concerning SMART and OIDC registration. I'm not sure best approach is admin has to register the form but instead turn on as feature in Globals. 

Also for the form, i'd follow a similar workflow as assigning users/providers where i'd also add, via ACL, new sections dealing specific to app permissions. This will make our routing validations easier in resource server.

Just talking out loud...

**Comment by adunsulag** [comment]
@sjpadgett As far as enabling SMART, I think having it on/off is fine in Globals.  I think what @vishnuyar is saying is we need a place where people can go to register their apps and manage the permissions.  I don't care if that is in Globals or the Forms system.  I'm not familiar with how the forms work so if its a good place to manage the apps that's fine with me, on the surface it sounds strange to treat a form as a webapp, but again I'm not familiar with how the forms functions.  

I'm thinking what it would look like for a clinic to try and add a SMART app and I think it would be a little strange to have to go to a Forms administration page in order to enable a SMART app.

Have you already done any work on building this UI @vishnuyar?  I'm going to first start working heavily on handling the scope resolution for permission checks and ACLs with the FHIR resources.  So if you are wanting to take on some of that UX of registering / managing the apps that's fine with me.  We are hoping to have all of the UX/UI and the SMART on FHIR integration pieces done by the end of the month.

**Comment by sjpadgett** [comment]
> I am thinking, we will need a create a UI for registering SMART App's.

I'm not sure where we are on an EMR app for the original intent of this issue but, I was responding to the above.
So besides having a dynamic registration, it behooves us to have app registration built into an administrators prevue.
Perhaps we need to separate that into a separate issue as we seem to be taking three different paths in this issue thread!
Or am I just showing my ignorance of the SMART implementation?

**Comment by sjpadgett** [comment]
> Create a Form for SMART APP Launch
> This can be installed from Administration -->Forms --> Forms Administration

To be clear, this is best served from forms registered forms but we need a clearer description of what a provider would use the form for and when/why a form would implement a SMART launch. e.g soliciting a FHIR QuestionnaireReponse via a launch of our EasiPro module for a portal request to patient(again just an example workflow). Too me, this form would create a SMART launch to a different server. Though to demonstrate a ONC EMR launch then it would be helpful to understand the purpose of the form.

I'm thinking we should have a conference call so we all are on the same page. Perhaps @stephenwaite could help us make that happen from openemrs normal meeting method.

**Comment by stephenwaite** [comment]
@bradymiller donates this service for any openemr related needs:
https://www.open-emr.org/wiki/index.php/OpenEMR_Foundation_Monthly_Conference_Call_Details

**Comment by sjpadgett** [comment]
Okay, guess i'll just figure out how to conf from my phone. Can't be too hard and I am an engineer!:)

**Comment by stephenwaite** [comment]
oh, i mean feel free, it's wide open for use, just follow that guide

**Comment by sjpadgett** [comment]
Following instructions is just not my thing! I am a rebel after all!:)

Okay enough of my musings... thanks Stephen W.

**Comment by vishnuyar** [comment]
@sjpadgett , I agree there are two many pieces which needs to be pieced to have a nice SMART App invocation. may be a design document approach is better before embarking on this. (http://hl7.org/fhir/smart-app-launch/#smart-authorization-and-resource-retrieval)
@adunsulag , I haven't started any work on the UI part. My thinking is based on how blue button(https://sandbox.bluebutton.cms.gov/v1/accounts/mfa/login) . Here the scope permissions for the SMART App can be granted/restricted.

We also need to think about CDS hooks (https://cds-hooks.org/) for OpenEMR
I am envisaging OpenEMR has a storage of Medical Records, and all supplementary actions  such as Pre-Authorisation permission, Growth Charts etc. to be taken by SMART Apps

**Comment by sjpadgett** [comment]
Okay guys, i'm thinking maybe i'll just leave this to you as i'm not taking much effort to learn the SMART if I don't have to(too much other stuff i'm working).
However, I kind of have an opinion concerning registration. How we register apps is entirely up to us and I see two forms. 
- Dynamic where Smart can be added as an application type; smart, smart-public or whatever we choose to indicate a clients role. This would be the fastest way to move forward with testing.
- Manual where provider determines access. New form in our normal authorization admin workflow i.e Users.

We also need to create the ACL groups and sections to support our various claims/permissions in our routes. Perhaps one of our more ACL knowledgeable admins will help here. lol, yeah that's you Brady! 

@vishnuyar I always envisioned that we'd move more and more towards offloading some of provider workflows to apps. Especially those practice discipline specific features. Even portal should go away as a patient option. So, go wild with whatever would be challenging for you.
I will note however that at some point i'd like us to implement adaptive questionnaires and responses for CDR scoring.

I'm off to do the client grant type in server. At least it's not the password grant anymore:)

**Comment by sreekanthmettles** [comment]
Hi there, we have a smart on FHIR application which can help in Prior Authorization. We have tested it with Cerner. We are on onboarding to their App Gallery. We can test the same on OpenEmr we get a methodology to launch our app.

**Comment by sjpadgett** [comment]
Hi @sreekanthmettles
How do you register your app?

**Comment by sreekanthmettles** [comment]
In cerner they have a registration page for the app vendors. We need to mention the launch URL, the redirect URL and the scopes when registering the app.

**Comment by sjpadgett** [comment]
If they had dynamic registration, would you see value with that or manual is preferred?

**Comment by sjpadgett** [comment]
Okay, well i'd like to see both guys...

**Comment by sreekanthmettles** [comment]
> If they had dynamic registration, would you see value with that or manual is preferred?

Cerner being a commercial proprietary software, they may be needing registration for multiple reasons. Not all of them may be technical. However in case of OpenEmr, I would prefer a Dynamic registration.

**Comment by sjpadgett** [comment]
Could you be so kind as to list what scopes your app requires/wants?

**Comment by sjpadgett** [comment]
I'm adding the smart requirements to the existing oidc dynamic registration. So, this will be in place for y'all. Prob tonight or tomorrow.

**Comment by sreekanthmettles** [comment]
We extract the medical record of the patient as a part of our Prior Auth request process for supporting documentation. So we use most of the resources that are defined by USCDI. We need all these resources in "Read" mode. For "write", we need QuestionnaireResponse, Task and ClaimResponse. However we can get around if the write options are not available...

**Comment by adunsulag** [comment]
@vishnuyar My latest changes in my pull request fixes the CORS problem.  Long term I'd like to match the CORS request based upon the client app registration to the domain origin, but for now this just block grants '*' to anyone on OPTIONS.  If you have a valid token it also grants CORS request on standard REST verbs (PUT, POST,GET, etc).

I've got the beginnings of the launch/patient working to a point that I was able to get from SMART app launch to making patient requests with FHIR using the SMART js api.  Right now the server just returns the first Patient found in the system for testing purposes.  We still need to build a patient selector widget as part of the OAUTH2 authorization protocol for administrators.  For patients I need to set it to use the currently logged in user as the selected patient.  So its a start and I thought I'd give y'all an update on where I'm at.

I also put in some more detailed logging for the OAUTH2 flow.
